### ISSUE TITLE
std.ArrayListUnmanaged: save 1 pointer-sized integer of memory overhead by making capacity inferred based on item count

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#5815]
Status quo:

```zig
items: []T,
capacity: usize,
```

This is 24 bytes on 64-bit architectures. If we instead made it:

```zig
items: []T,
```

Then it would be 16, the same size as any slice. The capacity would be inferred based on `items.len`.

The downside is that we would be giving up the ability to do `shrinkRetainingCapacity`. A shrink that caused the inferred capacity to change would need to actually use the allocator to give up the space. We would also be giving up the ability to accept overallocations from the underlying allocator.

The main use case for this would be to save on memory usage when there are many ArrayList in memory.

**Comment by daurnimator** [comment]
> The capacity would be inferred based on `items.len`.

howso?

**Comment by andrewrk** [comment]
For example, one implementation would be:

```zig
fn capacity(items_len: usize) usize {
    if (items_len == 0) return 0;
    var n: usize = 1;
    while (n < items_len)
        n *= 2;
    return n;
}
```

I'm sure there's a mathy bit fiddly way to efficiently compute capacity based on any growth factor we want.

**Comment by tgschultz** [comment]
This makes the assumption that the capacity available has been determined by the growth behavior. This is not the case when fromOwnedSlice is employed.

**Comment by zzyxyzz** [comment]
This seems like a lot of computational overhead to save a word of memory. Two problems come to mind:

1. The capacity can only be efficiently inferred for power-of-two sizes, with other growth factors requiring a loop or a table lookup. However, doubling capacity each time would totally negate the memory saved by not storing capacity explicitly.
2. Memory needs to be returned to the allocator when the array shrinks past a boundary, as you already pointed out. This is not only not good on average, but also leads to a very bad worst case, when an element is repeatedly pushed and popped across a size boundary, which may require a complete copy of the array, making insertion and deletion potentially O(n).

In short, this seems like a hack that gives up a lot to gain rather little. If memory overhead from many ArrayLists is a concern, it would probably be better to introduce a SmallArrayList that uses smaller integers for size and capacity and reduces the growth factor to 1.3 or something.

**Comment by zzyxyzz** [comment]
Correction: My second objection can be mitigated by storing an extra bit (e.g. in a pointer tag) that indicates that the size has shrunk past a boundary, but the array has not been reallocated yet. This way, the array would only be reallocated when the size shrinks past the *next* boundary. This would restore O(1) amortized insertion/deletion, but I still think that all of the additional juggling is not worth the 8 bytes per array.
